### PR TITLE
HOT-1617 - Saving Indicators

### PR DESCRIPTION
**Pull request by wmitchell-cambria**
### Jira Story

- [Display loading and saving animations HOT-1617](https://osi-cwds.atlassian.net/browse/HOT-1617)

## Description

This takes the LoadingButton (now called SavingButton) that I merged previously and uses it on each card of Hotline.

Sorry its so big...aren't you glad I merged that earlier PR already?

### How cards transition modes

When a card is in Edit mode, clicking save triggers a 'saveCard' action. ScreeningPageReducer listens for these actions and sets the corresponding card mode to 'saving'. The mode is already selected to deal with edit/show, and so saving comes free. The components listens for this and renders its saving mode. Once the saving is complete, a SAVE_SCREENING_COMPLETE action is fired. The same reducer listens for this and marks any saving card to show mode.

Lastly, the cards themselves do a little extra legwork as well. Because the transition to show mode is delayed, so is the hash navigation to set the scroll to the top of the card. Instead of happening immediately on save, the cards listen for transitions between saving mode and show mode (for saving) and between edit mode and show mode (for canceling). This is a little more complex than ideal, but then again, resetting the scroll is inherently impure.

### Saving mode

Saving mode basically looks like edit mode, but with a saving button. We assume for now that the user is not saving multiple cards at the same time (just like we already assume today).

### Other inclusions

Some feature tests were broken, but happened to be passing until these changes. Their exceptions were being silently handled by sagas, and the page sorta worked OK enough for the tests to pass. So, some unrelated feature tests fixes around allegations are included here.

Generating Baby Doe (for safely surrendered baby reports) got a little refactor as well. Previously, it was firing a saveCard action for allegations. This had the side effect of displaying a saving indicator unexpectedly and navigation halfway down the page on completion. I refactored the saveScreeningCard saga to share code directly instead.

Some components became PureComponents. Mostly the cards, because they needed a lifecycle method. PureComponents should actually be more performant for these than SFCs, since they will re-render less often.

Sleeps. In order to test the saving spinner states in feature tests, we had to include an artificial delay in a handful of our stubbed API calls. We searched long and far for a better way, and this was the cleanest approach we could find. It only applies to tests that care about saving state, so the impact on test run time is minimal.

## Tests
- [x] I have included unit tests 
- [x] I have included feature tests 
- [ ] I have included other tests 
- [ ] I have NOT included tests 

## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have ran all tests for the project.
- [x] I have ran lint/rubocop check for the changed/new files.
- [x] My code is in a stable state ready to be deployable, but not necessarily complete.
- [x] I promise on my honor as a CWDS developer to ensure I don't break the build, to check Lint/Rubocop, use best practices, to leave the code in better shape than I found it, and to have a good day.

### CODE REVIEW

**Comment by wmitchell-cambria** [comment]
Jenkins retest this please.

**Comment by wmitchell-cambria** [comment]
I'm still trying to figure out that one feature test. It consistently fails on Jenkins and passes locally.

**Comment by wmitchell-cambria** [comment]
Retest this please.

**Comment by money** [comment]
Tested locally using acceptance test to successfully submit referral on Chrome, Firefox, and Safari on Mac.

**Comment by wmitchell-cambria** [comment]
Weird results that seem unrelated to the code. Investigating, but in the meantime retest this please.

File: spec/features/screening/narrative_spec.rb [hunk 1]
```diff
@@ -80,18 +80,22 @@
     existing_screening[:report_narrative] = 'Trying to fill in with changes'
     stub_request(
       :put, ferb_api_url(FerbRoutes.intake_screening_path(existing_screening[:id]))
-    ).and_return(json_body(existing_screening.to_json))
+    ).and_return(proc {
+      sleep 2
```
Comment (pokharelmanish) [inline spec/features/screening/narrative_spec.rb]: never seen this before `sleep`?

Comment (wmitchell-cambria) [reply spec/features/screening/narrative_spec.rb]: Yeah, we need these in order to make sure that the 'Saving...' state is present long enough for Capybara to see it. We tried putting expectations in different places or structuring them differently, but the sleep was the only approach we could get working.

File: spec/features/screening/cross_reports_spec.rb [hunk 2]
```diff
@@ -70,6 +74,10 @@
         )
       )
     ).to have_been_made
+
+    within '#cross-report-card.show', wait: 4 do
```
Comment (NagaPradeepGajula-osi) [inline spec/features/screening/cross_reports_spec.rb]: same here is this to account for the loading time? what if it takes more than 4?

File: spec/features/screening/decision_spec.rb [hunk 3]
```diff
@@ -122,8 +125,9 @@
       select 'Mark as Sensitive', from: 'Access Restrictions'
       fill_in 'Restrictions Rationale', with: 'Someone in this screening has sensitive information'
       click_button 'Save'
+      expect(page).to have_button('Saving', disabled: true)
     end
-    within '#decision-card.show' do
+    within '#decision-card.show', wait: 4 do
```
Comment (NagaPradeepGajula-osi) [inline spec/features/screening/decision_spec.rb]: same as above

File: spec/features/screening/incident_information_spec.rb [hunk 4]
```diff
@@ -80,20 +80,24 @@
...
+      expect(page).to have_button('Saving', disabled: true)
     end
 
     expect(
       a_request(:put, ferb_api_url(FerbRoutes.intake_screening_path(existing_screening[:id])))
         .with(body: hash_including(existing_screening))
     ).to have_been_made
 
-    within '#incident-information-card.show' do
+    within '#incident-information-card.show', wait: 4 do
```
Comment (NagaPradeepGajula-osi) [inline spec/features/screening/incident_information_spec.rb]: same here

File: spec/features/screening/narrative_spec.rb [hunk 5]
```diff
@@ -117,20 +121,24 @@
...
+      expect(page).to have_button('Saving', disabled: true)
     end
 
     expect(
       a_request(:put, ferb_api_url(FerbRoutes.intake_screening_path(existing_screening[:id])))
         .with(body: hash_including(existing_screening))
     ).to have_been_made
 
-    within '#narrative-card.show' do
+    within '#narrative-card.show', wait: 4 do
```
Comment (NagaPradeepGajula-osi) [inline spec/features/screening/narrative_spec.rb]: 2 more here

File: spec/javascripts/selectors/screening/screeningPageSelectorsSpec.js [hunk 6]
```diff
@@ -10,30 +11,37 @@ import {fromJS} from 'immutable'
 describe('screeningPageSelectors', () => {
   describe('getCardModeValueSelector', () => {
     it('returns the current mode for a given card', () => {
-      const screeningPage = {cards: {'some-card': 'show'}}
+      const screeningPage = {cards: {'some-card': 'limbo'}}
       const state = fromJS({screeningPage})
-      expect(getCardModeValueSelector(state, 'some-card')).toEqual('show')
+      expect(getCardModeValueSelector(state, 'some-card')).toEqual('limbo')
```
[the remark this answers is not in the document]
Comment (NagaPradeepGajula-osi) [reply spec/javascripts/selectors/screening/screeningPageSelectorsSpec.js]: 👍

File: spec/features/screening/allegations/edit_allegations_spec.rb [hunk 7]
```diff
@@ -310,7 +310,10 @@
     screening[:allegations] << persisted_allegation
 
     stub_request(:put, ferb_api_url(FerbRoutes.intake_screening_path(screening[:id])))
-      .and_return(json_body(screening.to_json, status: 200))
+      .and_return(proc {
+        sleep 2
```
[the remark this answers is not in the document]
Comment (NagaPradeepGajula-osi) [reply spec/features/screening/allegations/edit_allegations_spec.rb]: 👍

**Pull request merged**